{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Installation notes \n",
    "\n",
    "##### Prerequisites \n",
    "Python v3 \\\n",
    "Jupyter notebook \n",
    "\n",
    "##### Python packages to run the code\n",
    "pip install networkx \\\n",
    "pip install matplotlib "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gfa parser to write the gfa file to dictiory and add repeats after confirmation\n",
    "#the first step in the code \n",
    "from collections import defaultdict\n",
    "\n",
    "class gfa_parser:\n",
    "    '''\n",
    "    Read the spades output in GFA v1.0 graph to genrate \n",
    "    nodes and edges. \n",
    "    \n",
    "    #Class initiation\n",
    "    x=gfa_parser(\"chrom2.gfa\")\n",
    "    \n",
    "    #calling a specific function within the class\n",
    "    print(x.gfa_parser(\"chrom2.gfa\"))\n",
    "    #prints all the nodes and edges from the gfa \n",
    "    \n",
    "    '''\n",
    "    def __init__(self, gfa):\n",
    "        self.gfa=gfa\n",
    "        \n",
    "    \n",
    "    def gfa_parser(self, gfa):\n",
    "        #reading the gfa file to a new variable\n",
    "        g1=open(gfa, \"r\")\n",
    "        nodes=[]\n",
    "        edges=[]\n",
    "        coverage={}\n",
    "        sequences={}\n",
    "        for line in g1:\n",
    "            if (line.startswith(\"L\")):\n",
    "                link=line.split('\\t')\n",
    "                edges.append([link[1],link[3]])\n",
    "            \n",
    "            elif (line.startswith(\"S\")):\n",
    "                info=line.split('\\t')\n",
    "                seq=float(info[1])\n",
    "                sequences[seq]=info[2]\n",
    "                cov=info[-1].split(\":\")[-1]\n",
    "                cov=float(cov.strip(\"\\n\"))\n",
    "                if (cov>2.0):\n",
    "                    coverage[info[1]]=cov\n",
    "        \n",
    "        #user input\n",
    "        repeats=[]\n",
    "        for f in repeats:\n",
    "            cv=coverage.get(f)\n",
    "            cv=cv-1\n",
    "            temp=f\n",
    "            for index in range (int(cv)):\n",
    "                rename=str(f)+\"r\"+str(index)\n",
    "                edges.append([temp,rename])\n",
    "                temp=rename\n",
    "                \n",
    "        return(edges, coverage, sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notes\n",
    "\n",
    "***Vibrio splendidus* genome Chromosome 1 contains 7 rRNA genes and 97 tRNA genes, while chromosome 2 contains 1 rRNA gene and 17 tRNA genes**\n",
    "\n",
    "Below includes order of the rrn operon in chromosome1 and chromosome2. Confirmed  these sequences using BLAST as well, there are several nanopore reads with these sequences - Example 6de21c81-fa59-4e88-9fe0-b1d585bfa675 has 10 of the below sequences and 99cce163-3455-46f7-8887-5483343d16b7.\n",
    "\n",
    "|Node |coverage |length | num of hits before|num of hits after| Annotation |\n",
    "|---|-----|-----|---------|-----|-----|\n",
    "|25| 3.91| 85bp | 6 |11| 5S ribosomal sequence| \n",
    "|26| 11.96| 81bp  |6 |11| 5S ribosomal sequence |\n",
    "|32| 13.16 | 57bp | 5| 9| no associated gene/protein |\n",
    "|9 | 8.45 | 2941bp | 4| 9| 23S ribosomal sequence|\n",
    "|29| 2.91| 61bp| 5 | 9| 16S-23S ribosomal RNA intergenic spacer|\n",
    "|30 | 8.18| 57bp |5| 9| 16S-23S ribosomal RNA intergenic spacer|\n",
    "|18| 11.85| 168bp |4 | 9| 23S ribosomal sequence|\n",
    "|16| 4.89| 276bp | 6 | 13| 16S-23S ribosomal RNA intergenic spacer|\n",
    "|11 | 14.56 |1604bp | 4 | 9| 16S ribosomal sequence|\n",
    "|31| 11.07| 57bp | 5 | 10|no associated gene/protein |\n",
    "|21| 9.50| 152bp |6| 11| no associated gene/protein|\n",
    "|24| 5.85| 140bp | 6 | 11| no associated gene/protein| \n",
    "|27| 5.04| 75bp | 5 | 10| no associated gene/protein|\n",
    "\n",
    "Conclusion - If the hits contained 1 23S and 1 16S sequence, that was counted as an rrn operon. In total chromosom1 has 8 operons and chromsome2 has 1 operon.\n",
    "\n",
    "|Node |coverage |length | num of hits before|num of hits after| Annotation |Likely repeat sequence |\n",
    "|---|-----|-----|---------|-----|-----|----|\n",
    "|22| 2.39| 144bp |5 | 10| tRNA sequence|multiple places|\n",
    "|17| 3.09| 205bp |4 | 7| tRNA sequence|multiple places|\n",
    "\n",
    "Conclusion - The tRNA sequences align at multiple places - left this as is.\n",
    "\n",
    "|Node |coverage |length | num of hits before|num of hits after| Annotation |Likely repeat sequence | nanopore reads |\n",
    "|---|-----|-----|---------|-----|-----|----|---|\n",
    "|28| 6.73| 70bp| 1 | 1| 16S-23S ribosomal RNA intergenic spacer| repeat|6e5d2ec6-8da1-4d65-b4b3-7ca64210c18d, 8325866e-2b95-405e-9a53-f1b35696d56d, 930ddd43-974e-4c91-978b-5b0eaccc41ad (100% pid and 70bp in length)|\n",
    "|12 (self loop)| 41.66| 336bp | 2| 43| hypothetical protein |repeat|aa3f95b9-4aa9-4639-8e56-5a53cd42e95d, da702a33-0ac2-4993-a19d-539d7206489d, bc276ee2-23d6-45ef-a9fb-aca9fdb72e2c (top three reads, but each of these reads show that the 12 seqeunces is repeated on the same nanopore read)|\n",
    "|23| 10.54| 142bp |1 | 1| no associated gene/protein |repeat|aae08399-414f-4b89-943c-d997cd1e51ec (best hit), multiple reads but doesnt show evidence of repeats like 12|\n",
    "|33| 7.59| 42bp | 2 | 2| no associated gene/protein|repeat|aae08399-414f-4b89-943c-d997cd1e51ec,06008824-bf67-4a6e-ad3b-6c146a5a629c, 8325866e-2b95-405e-9a53-f1b35696d56d, ed3cf1af-9eb3-463c-8360-735a5de87e8a, 1cef8308-f6d3-4efb-b787-247161ca4418, 97c0acbd-e6d8-4884-9a70-8ac6481fb784 (100%pid and 42bp in length)|\n",
    "\n",
    "Conclusion - Since node 12 is definitely a repeat, added this repeat to the seqeuence. \n",
    "\n",
    "|Node |coverage |length | num of hits before|num of hits after| Annotation |\n",
    "|---|-----|-----|---------|-----|-----|\n",
    "|36| 8.12| 27bp | 0 | 0| -|\n",
    "|37| 3.16| 21bp | 0 | 0| - |\n",
    "|38| 2.84 | 15bp | 0 | 0| - |\n",
    "|39| 4.58| 7bp | 0 | 0| -|\n",
    "|41 | 5.86| 6bp| 0| 0| -| \n",
    "\n",
    "Conclusion - ignored the above sequences \n",
    "\n",
    "**This genome has 8 rrn operons in chrom1 and 1rrn operon in chrom2.**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Python3 program to find the maximum depth of tree from a given node\n",
    "# counts the depth of the tree from the left tree and right tree\n",
    "# Python3 program to find the maximum depth of tree \n",
    "import networkx as nx\n",
    "# A binary tree node \n",
    "def maxDepth (node, dfs, graph): \n",
    "    if node is None: \n",
    "        return (0)\n",
    "    else : \n",
    "        # Compute the depth of each subtree \n",
    "        n=list(G.nodes())[0]\n",
    "        value=list(dfs[node])\n",
    "        left_tree=0\n",
    "        right_tree=0\n",
    "        left=value[0]\n",
    "        right=value[1]\n",
    "        while (left is not None):\n",
    "            try:\n",
    "                left=dfs[left][0]\n",
    "                left_tree=left_tree+1\n",
    "            except KeyError:\n",
    "                break\n",
    "        while (right is not None):\n",
    "            try:\n",
    "                right=dfs[right][0]\n",
    "                right_tree=right_tree+1\n",
    "            except KeyError:\n",
    "                break\n",
    "        \n",
    "    return (left_tree+1, right_tree+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#if the node has two branches\n",
    "# 1) check the depth of the right and the left branch\n",
    "# 2) traverse through the branch that is the shortest then traverse back to the node that had two branches back to the next branch\n",
    "# 3) also checks to make sure the first node remains to have only 1 neighbor\n",
    "def adding_nodes_with_two_branches(key, successors, old_graph, new_graph):\n",
    "    d=maxDepth(key, successors, old_graph)\n",
    "    left_depth=d[0]\n",
    "    right_depth=d[1]\n",
    "    \n",
    "    #if the maximum depth to the left tree and right tree are both 1\n",
    "    if (right_depth==1) and (left_depth==1):\n",
    "        new_graph=check_before_adding(key, val[0], new_graph)\n",
    "        #getting the last node\n",
    "        last_node=get_last_edge(new_graph)[1]\n",
    "        #tracking back\n",
    "        new_graph=check_before_adding(last_node, key, new_graph)\n",
    "        new_graph=check_before_adding(key, val[1], new_graph)\n",
    "                    \n",
    "    #if the left depth is less than the right depth\n",
    "    elif (left_depth<right_depth):    \n",
    "        #print (\"left depth is less than right depth\", left_depth, right_depth)\n",
    "        if (left_depth==1):\n",
    "            new_graph=check_before_adding(key, val[0], new_graph)\n",
    "            #getting the last node\n",
    "            last_node=get_last_edge(new_graph)[1]\n",
    "            #tracing back\n",
    "            new_graph=check_before_adding(last_node, key, new_graph)\n",
    "            new_graph=check_before_adding(key, val[1], new_graph)\n",
    "        else:\n",
    "            new_graph=check_before_adding(key, val[0], new_graph)\n",
    "            successors[key]=(val[1])\n",
    "\n",
    "    #if the right depth is less than the left depth\n",
    "    elif (left_depth>right_depth):\n",
    "        #print (\"left depth is more than right depth\", left_depth, right_depth)\n",
    "        if (right_depth==1):\n",
    "            new_graph=check_before_adding(key, val[1], new_graph)\n",
    "            #getting the last node\n",
    "            last_node=get_last_edge(new_graph)[1]\n",
    "            #tracing back\n",
    "            new_graph=check_before_adding(last_node, key, new_graph)\n",
    "            new_graph=check_before_adding(key, val[0], new_graph)\n",
    "        else:\n",
    "            new_graph=check_before_adding(key,val[1], new_graph)\n",
    "            successors[key]=(val[0])\n",
    "            \n",
    "    return (successors, new_graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting the last edge and nodes added to the new tree\n",
    "def get_last_edge(new_graph):\n",
    "    edges=list(new_graph.edges())\n",
    "    last_edge=edges[-1]\n",
    "    last_node=last_edge[1].split('_')[0]\n",
    "    prev_node=last_edge[0].split('_')[0]\n",
    "    return (last_edge, last_node, prev_node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# this function only renames the node since it already exists in the \n",
    "#   new graph \n",
    "# assumes that the node was already checked to confirm that \n",
    "#   the node is already added to the new graph \n",
    "def rename_node(node, new_graph):\n",
    "    nodes=new_graph.nodes()\n",
    "    count=1\n",
    "    new_node=node+\"_\"+str(count)\n",
    "    if (new_node in nodes):\n",
    "        ngbr=len(list(new_graph.neighbors(new_node)))\n",
    "        if (ngbr==1):\n",
    "            new_node=new_node\n",
    "        elif (ngbr==2):\n",
    "            count=count+1\n",
    "            new_node=node+\"_\"+str(count)\n",
    "    \n",
    "    return (new_node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#before adding a  new edge, check if the nodes in the edge are \n",
    "#  1) first that the node exists or not in the new_graph \n",
    "#  2) check if the node neighbors are 2 or 1\n",
    "#  3) node1 and node2 is not the first node in the graph\n",
    "def check_before_adding(node1, node2, new_graph):\n",
    "    \n",
    "    #getting the first node in the new graph\n",
    "    n=list(G.nodes())[0]\n",
    "        \n",
    "    if (node1 not in new_graph.nodes()) and (node2 not in new_graph.nodes()):\n",
    "        new_graph.add_edge(node1, node2)\n",
    "\n",
    "    elif (node1 in new_graph.nodes()) and (node2 not in new_graph.nodes()):\n",
    "        n1_ngbrs=len(list(new_graph.neighbors(node1)))\n",
    "        if (n1_ngbrs==1) and (node1 !=n):\n",
    "            new_graph.add_edge(node1, node2)\n",
    "        elif (n1_ngbrs==1) and (node1 ==n):\n",
    "            x=rename_node(node1, new_graph)\n",
    "            new_graph.add_edge(x, node2)\n",
    "        elif (n1_ngbrs==2):\n",
    "            x=rename_node(node1, new_graph)\n",
    "            new_graph.add_edge(x, node2)\n",
    "\n",
    "    elif(node1 not in new_graph.nodes()) and (node2 in new_graph.nodes()):\n",
    "        n2_ngbrs=len(list(new_graph.neighbors(node2)))\n",
    "        if (n2_ngbrs==1) and (node2 !=n):\n",
    "            new_graph.add_edge(node1, node2)\n",
    "        elif (n2_ngbrs==1) and (node2 ==n):\n",
    "            x=rename_node(node2, new_graph)\n",
    "            new_graph.add_edge(node1, x)\n",
    "        elif (n2_ngbrs==2):\n",
    "            x=rename_node(node2, new_graph)\n",
    "            new_graph.add_edge(node1, x)\n",
    "\n",
    "    elif (node1 in new_graph.nodes()) and (node2 in new_graph.nodes()):\n",
    "        n1_ngbrs=len(list(new_graph.neighbors(node1)))\n",
    "        n2_ngbrs=len(list(new_graph.neighbors(node2)))\n",
    "        if (n1_ngbrs==1) and (n2_ngbrs==1):\n",
    "            if (node1 ==n) and (node2 !=n):\n",
    "                x=rename_node(node1, new_graph)\n",
    "                new_graph.add_edge(x,node2)\n",
    "            elif (node1 !=n) and (node2 ==n):\n",
    "                x=rename_node(node2, new_graph)\n",
    "                new_graph.add_edge(node1, x)\n",
    "            else:\n",
    "                new_graph.add_edge(node1, node2)\n",
    "\n",
    "        elif (n1_ngbrs==2) and (n2_ngbrs==1):\n",
    "            if (node2 !=n):\n",
    "                x=rename_node(node1, new_graph)\n",
    "                new_graph.add_edge(x, node2)\n",
    "            else:\n",
    "                k=rename_node(node1, new_graph)\n",
    "                v=rename_node(node2, new_graph)\n",
    "                new_graph.add_edge(k,v)\n",
    "    \n",
    "        elif (n1_ngbrs==1) and (n2_ngbrs==2):\n",
    "            if (node1 !=n):\n",
    "                x=rename_node(node2, new_graph)\n",
    "                new_graph.add_edge(node1,x)\n",
    "            else:\n",
    "                k=rename_node(node1, new_graph)\n",
    "                v=rename_node(node2, new_graph)\n",
    "                new_graph.add_edge(k,v)\n",
    "        else:\n",
    "            k=rename_node(node1, new_graph)\n",
    "            v=rename_node(node2, new_graph)\n",
    "            new_graph.add_edge(k,v)\n",
    "    return(new_graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# finding the paths from the last node to the first node and selecting the \n",
    "# shortest path\n",
    "# using the dictionary to retrieve the neighbors\n",
    "def path_to_first_node(last_node, first_node, old_graph):\n",
    "    #getting the neighbors of the last node and working on adding the nodes to the graph accordingly\n",
    "    last_node=last_node.split(\"_\")[0]\n",
    "    last_node=last_node.split(\"r\")[0]\n",
    "    nbrs=list(old_graph.neighbors(last_node))\n",
    "    left_path=list(nx.shortest_path(old_graph, source=nbrs[0], target=first_node))\n",
    "    right_path=list(nx.shortest_path(old_graph, source=nbrs[1], target=first_node))\n",
    "\n",
    "    return (left_path, right_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#traversing through the two shortest paths from the last node to the key \n",
    "def shortest_path_to_first_node(left_path, right_path, new_graph):\n",
    "                                \n",
    "    #getting the last_node added to the new_graph \n",
    "    last_node=list(new_graph.nodes())[-1]\n",
    "    \n",
    "    #traversing the shortest path back\n",
    "    if (len(left_path)==len(right_path)):\n",
    "        for f in range(len(left_path)):\n",
    "            if (f==0):\n",
    "                new_graph=check_before_adding(last_node, left_path[f], new_graph)\n",
    "            else:\n",
    "                new_graph=check_before_adding(left_path[f-1], left_path[f], new_graph)\n",
    "                \n",
    "    elif (len(left_path)<len(right_path)):\n",
    "        for f in range(len(left_path)):\n",
    "            if (f==0):\n",
    "                new_graph=check_before_adding(last_node, left_path[f], new_graph)\n",
    "            else:\n",
    "                new_graph=check_before_adding(left_path[f-1], left_path[f], new_graph)\n",
    "                \n",
    "    elif (len(left_path)>len(right_path)):\n",
    "        for f in range(len(right_path)):\n",
    "            if (f==0):\n",
    "                new_graph=check_before_adding(last_node, right_path[f], new_graph)\n",
    "            else:\n",
    "                new_graph=check_before_adding(right_path[f-1], right_path[f], new_graph)\n",
    "    \n",
    "    return (new_graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Writing out the complete graph as a fasta file. \n",
    "# the new complete graph is written out to \"new_graph\" variable, \n",
    "# but this holds only inforamtion about the node names and the sequence itself.\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "def graph_to_fasta(graph, name, sequences):\n",
    "    #made sure the nodes were added in the correct order\n",
    "    nodes=graph.nodes()\n",
    "    final_seq=\"\"\n",
    "    for f in nodes:\n",
    "        raw_node=f.split(\"_\")[0]\n",
    "        raw_node=raw_node.split(\"r\")[0]\n",
    "        #print (raw_node)\n",
    "        seq=sequences[int(raw_node)]\n",
    "        final_seq = final_seq+seq\n",
    "    final_seq=Seq(final_seq)\n",
    "    \n",
    "    #writing out the final seq as a record class\n",
    "    final = SeqRecord(final_seq, id=name, description=\"putting it together\")\n",
    "    return (final)\n",
    "    \n",
    "###############################################\n",
    "# Driver program\n",
    "##############################################\n",
    "#final=graph_to_fasta(new_graph, sequences)\n",
    "#print (final.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "key, val, count 2 ['12'] 0\n",
      "key, val, count 12 ['3', '12r0'] 1\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'12'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-de2b871e63fa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     73\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m                 \u001b[0;31m#print (\"adding edges with more than one path\")\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m                 \u001b[0mtwo_branches\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0madding_nodes_with_two_branches\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msuccessors\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mG\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_graph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m                 \u001b[0msuccessors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtwo_branches\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m                 \u001b[0mnew_graph\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtwo_branches\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-29-19b5ac9d5984>\u001b[0m in \u001b[0;36madding_nodes_with_two_branches\u001b[0;34m(key, successors, old_graph, new_graph)\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# 3) also checks to make sure the first node remains to have only 1 neighbor\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0madding_nodes_with_two_branches\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msuccessors\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mold_graph\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_graph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0md\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmaxDepth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msuccessors\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mold_graph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m     \u001b[0mleft_depth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0mright_depth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-35-ac431bf4a022>\u001b[0m in \u001b[0;36mmaxDepth\u001b[0;34m(node, dfs, graph)\u001b[0m\n\u001b[1;32m     11\u001b[0m         \u001b[0;31m# Compute the depth of each subtree\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mG\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnodes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m         \u001b[0mvalue\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdfs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnode\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m         \u001b[0mleft_tree\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m         \u001b[0mright_tree\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: '12'"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "\n",
    "#calling the gfa parser first to retrieve the data from gfa\n",
    "#user input\n",
    "input_name=\"chrom2.gfa\"\n",
    "x=gfa_parser(input_name)\n",
    "edges=x.gfa_parser(input_name)[0]\n",
    "coverage=x.gfa_parser(input_name)[1]\n",
    "sequences=x.gfa_parser(input_name)[2]\n",
    "\n",
    "#using networkx package to write the edges from gfa to graph\n",
    "G=nx.Graph()\n",
    "G.add_edges_from(edges)\n",
    "#getting the first node in the graph\n",
    "n=list(G.nodes())[0]\n",
    "#dfs traversal\n",
    "successors=nx.dfs_successors(G,n)\n",
    "length=len(successors)\n",
    "print (successors)\n",
    "\n",
    "#traversing the old graph to write to a new graph, traverse through each node at least once\n",
    "new_graph=nx.Graph()\n",
    "count=0\n",
    "\n",
    "while (count < length):\n",
    "    keys=list(successors.keys())\n",
    "    key=keys[0]\n",
    "    val=successors[key]\n",
    "    print (\"key, val, count\",key, val, count)\n",
    "    #print (\"successors=\", successors)\n",
    "    \n",
    "    #adding the first edge if there is only branch down the tree\n",
    "    if (count==0) and(len(val)==1):\n",
    "        new_graph.add_edge(key, val[0]) \n",
    "        count=count+1 \n",
    "        successors.pop(key)\n",
    "    \n",
    "    #if the first node has two branches and picking the edge with the shortest depth\n",
    "    elif (count==0) and (len(val)!=1):\n",
    "        two_branches=adding_nodes_with_two_branches(key, successors, G, new_graph)\n",
    "        successors=two_branches[0]\n",
    "        count=count+1 \n",
    "        successors.pop(key)\n",
    "        \n",
    "    #adding the rest of the nodes \n",
    "    elif (count!=0):\n",
    "        #getting the last node\n",
    "        last_node=get_last_edge(new_graph)[1]\n",
    "        \n",
    "        #making sure the last node is the same as the next key\n",
    "        if (last_node==key):\n",
    "            \n",
    "            #if there is only one branch\n",
    "            if (len(val)==1):\n",
    "                new_graph=check_before_adding(key, val[0], new_graph)\n",
    "                count=count+1 \n",
    "                successors.pop(key)\n",
    "                \n",
    "            #if there are two branches\n",
    "            else:\n",
    "                #print (\"adding edges with more than one path\")\n",
    "                two_branches=adding_nodes_with_two_branches(key, successors, G, new_graph)\n",
    "                successors=two_branches[0]\n",
    "                new_graph=two_branches[1]\n",
    "                count=count+1 \n",
    "                successors.pop(key)\n",
    "   \n",
    "        \n",
    "        #if the last_node is not the same as the key, not in continuation \n",
    "        else:\n",
    "            \n",
    "            #finding the continuation node in the dict, if it exists \n",
    "            if (last_node in list(successors.keys())):\n",
    "                val=successors[last_node]\n",
    "                \n",
    "                #if there is only one branch from key\n",
    "                if (len(val)==1):\n",
    "                    new_graph=check_before_adding(last_node, val[0], new_graph)\n",
    "                    count=count+1 \n",
    "                    successors.pop(key)\n",
    "                \n",
    "                #if there is more than one branch down the tree\n",
    "                else:\n",
    "                    two_branches=adding_nodes_with_two_branches(key, successors, G, new_graph)\n",
    "                    successors=two_branches[0]\n",
    "                    new_graph=two_branches[1]\n",
    "                    count=count+1 \n",
    "                    successors.pop(key)\n",
    "                    \n",
    "            #finding the continuation node in the dict, if it doesn't exists \n",
    "            else:\n",
    "                #finding the shortest path to the key from the last node\n",
    "                path=path_to_first_node(last_node, key, G)\n",
    "                left_path=path[0]\n",
    "                right_path=path[1]\n",
    "                new_graph=shortest_path_to_first_node(left_path, right_path, new_graph)               \n",
    "\n",
    "    #print (\"count, length of keys, keys\", count, len(keys), keys)\n",
    "    #print (new_graph.edges())\n",
    "    print (\"successors=\", successors)\n",
    "    \n",
    "#adding last node\n",
    "last_node=get_last_edge(new_graph)[1]\n",
    "last_node=last_node.split('r')[0]\n",
    "path=path_to_first_node(last_node, n, G)\n",
    "left_path=path[0]\n",
    "right_path=path[1]\n",
    "new_graph=shortest_path_to_first_node(left_path, right_path, new_graph)               \n",
    "print (new_graph.edges())\n",
    "\n",
    "#print (new_graph.edges)\n",
    "nx.draw_networkx(new_graph)\n",
    "\n",
    "# writing the graph to fasta file \n",
    "final=graph_to_fasta(new_graph, input_name, sequences)\n",
    "SeqIO.write(final, \"ED252-chrom2-final.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Notes \n",
    "\n",
    "The last shell may have to be changed to functions as well.. Not sure how yet in a Jupyter notebook. If changed to a code, then add argparse python package. \n",
    "\n",
    "Replace the first line - name=\"chrom1.gfa\" and last line \"ED252-chrom1-final.fasta\". Replace the names for different files\n",
    "\n",
    "This generated two files, one for chromosome 1 and another for chromosome 2. In bash ran the command, \n",
    "    \n",
    "`cat ED252-chrom1-final.fasta ED252-chrom2-final.fasta >ED252-final.fasta`\n",
    "    \n",
    "   "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
